File: m02_ros2_communication/m02_ros2_with_py/m02_ros2_with_py/turtle_challenge.py
# !/usr/bin/env python3

# ------------------ Python Standard Libraries-------------------------------
import math

# ------------------ ROS2 Depedencies ----------------------------------------
import rclpy
from rclpy.node import Node

# -------------------- Turtlesim msg and srvs --------------------------------
from turtlesim.msg import Pose
from turtlesim.srv import Spawn, SetPen, Kill
from turtlesim.srv import TeleportAbsolute, TeleportRelative
from std_srvs.srv import Empty

# -------------- Class implementation for using turtlesim --------------------
class PlaygroundTurtle(Node):
    """
    Class oriented for turtlesim usage, oriented to generating the subscriptions
    and clients needed for using the functionalities of the turtles.
    
    Attributes
    ---
    cli : rclpy client
        Client implementation
        
    req : request
        Generic request for interaction between service and client
        
    res : response
        Generic request for interaction between service and client
        
    pub : rcply publisher
        Publisher implementation
        
    sub : rcply subscriber
        Subscriber implementation
    
    Methods
    ---
    clear_board():
        Clear screen of turtlesim
    
    spawn_turtle(x, y, theta, name):
        Create new turtle 
    
    kill_turtle(name):
        Delete given turtle
    
    teleport_abs_turtle(x, y, theta, name):
        Move turtle according world origin
    
    teleport_rel_turtle(v_x, v_theta, name):
        Move turtle relative to its tf with velocity
    
    set_pen_turtle(r, g, b, width, name):
        Set color and width trace of the turtle
    
    get_pose_turtle(name):
        Getter of pose of the given turtle
        
    save_pose_turtle(msg):
        Callback to store pose
    """
    def __init__(self):
        """
        Constructor that initialize the node with name 'playground_turtle' and
        defines the generic client, request, response, publisher and subscriber
        objects for connecting with turtlesim.
        """
        super().__init__('playground_turtle_py')
        self.cli = None
        self.req = None
        self.res = None
        self.pub = None
        self.sub = None
        self.pose = Pose()
    
    def clear_board(self):
        """
        Call service that clear the background of the turtlesim.
        
        Returns
        ---
            True if the response is received without errors.
        """
        self.cli = self.create_client(Empty, '/clear')
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger("Waiting turtlesim for action: Clear...")
        self.req = Empty.Request()
        self.res = self.cli.call_async(self.req)
        rclpy.spin_until_future_complete(self, self.res)
        return True
    
    def spawn_turtle(self, x, y, theta, name):
        """
        Call service that appears new turtles.
        
        Params
        ---
        x : float
            X position in the turtlesim map to spawn.
        y : float
            Y position in the turtlesim map to spawn.
        theta : float
            Orientation of the turtlesim to spawn (radians)
        name : String
            Name of the new turtle.
            
        Returns
        ---
            True if the response is received without errors.
        """
        self.cli = self.create_client(Spawn, '/spawn')
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger("Waiting turtlesim for action: Spawn...")
        self.req = Spawn.Request()
        self.req.x = float(x)
        self.req.y = float(y)
        self.req.theta = float(theta)
        self.req.name = str(name)
        self.res = self.cli.call_async(self.req)
        rclpy.spin_until_future_complete(self, self.res)
        return True
    
    def kill_turtle(self, name):
        """
        Call service that elimiates a certain turtle.
        
        Params
        ---
        name : String
            Name of the turtle to delete
            
        Returns
        ---
            True if the response is received without errors.
        """
        self.cli = self.create_client(Kill, '/kill')
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger("Waiting turtlesim for action: Kill...")
        self.req = Kill.Request()
        self.req.name = str(name)
        self.res = self.cli.call_async(self.req)
        rclpy.spin_until_future_complete(self, self.res)
        return True
    
    def teleport_abs_turtle(self, x, y, theta, name):
        """
        Call service that teleport a certain turtle (absolute position to
        turtlesim origin)
        
        Params
        ---
        x : float
            X position in the turtlesim map.
        y : float
            Y position in the turtlesim map.
        theta : float 
            Orientation of the turtle in space (radians).
        name : string
            Name of the turtle to mov
            
        Returns
        ---
            True if the response is received without errors.
        """
        self.cli = self.create_client(TeleportAbsolute, "/" + str(name) + 
                                        "/teleport_absolute")
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger("Waiting turtlesim for action: Teleport (Abs) " + 
                            str(name) + "...")
        self.req = TeleportAbsolute.Request()
        self.req.x = float(x)
        self.req.y = float(y)
        self.req.theta = float(theta)
        self.res = self.cli.call_async(self.req)
        rclpy.spin_until_future_complete(self, self.res)
        return True
    
    def teleport_rel_turtle(self, v_x, v_theta, name):
        """
        Call service that teleport a certain turtle (relative to the
        tranform of the same turtle)
        
        Params
        ---
        v_x : float
            Linear velocity to move oriented to x direction.
        v_theta : float
            Angular velocity to move oriented around z axis.
        name : string
            Name of the turtle to mov.
            
        Returns
        ---
            True if the response is received without errors.
        """
        self.cli = self.create_client(TeleportRelative, "/" + str(name) + 
                                        "/teleport_relative")
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger("Waiting turtlesim for action: Teleport (Rel) " + 
                            str(name) + "...")
        self.req = TeleportRelative.Request()
        self.req.linear = float(v_x)
        self.req.angulary = float(v_theta)
        self.res = self.cli.call_async(self.req)
        rclpy.spin_until_future_complete(self, self.res)
        return True
    
    def set_pen_turtle(self, r, g, b, width, name):
        """
        Set color and width of the trace of the given turtle.
        
        Params
        ---
        r : int
            Must be a value between 0 - 255 of red intensity
        g : int
            Must be a value between 0 - 255 of green intensity
        b : int
            Must be a value between 0 - 255 of blue intensity
        width : int
            Specify width of the line trace
        name : string
            Name of the turtle to change trace color
            
        Returns
        ---
            True if the response is received without errors.
        """
        self.cli = self.create_client(SetPen, "/" + name + "/set_pen")
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger("Waiting turtlesim for action: Set Pen " + 
                            str(name) + "...")
        self.req = SetPen.Request()
        self.req.r = int(r)
        self.req.g = int(g)
        self.req.b = int(b)
        self.req.width = int(width)
        self.res = self.cli.call_async(self.req)
        rclpy.spin_until_future_complete(self, self.res)
        return True
        
    def get_pose_turtle(self, name):
        """
        Getter of the pose of the turtle by using pose subscription via 
        callback
        
        Params
        ---
        name : string
            Name of the turtle to obtain the pose
        """
        self.sub = self.create_subscription(Pose, "/" + str(name) + "/pose", 
            save_pos_turtle, 10)
        
    def save_pos_turtle(self, msg):
        """
        Callback for getting the pose of the given turtle.
        
        Params
        ---
        msg : Turtle Pose
            Pose obtained via message of the turtle
        """
        self.pose = msg
        
    
def main():
    """
    Main oriented to test the playground functions that interacts with 
    turtlesim.
    """
    # Initialize ROS Client Library for Python
    rclpy.init()
    
    # Define name and instance class
    name1 = "dan_turtle"
    my_turtle = PlaygroundTurtle()
    
    # Start playing with turtlesim
    my_turtle.clear_board()
    my_turtle.spawn_turtle(3, 3, 0, name1)
    my_turtle.teleport_abs_turtle(7, 3, 3*math.pi/2, name1)
    my_turtle.set_pen_turtle(255, 0, 0, 2, name1)
    my_turtle.teleport_abs_turtle(7, 7, -math.pi, name1)
    my_turtle.set_pen_turtle(0, 255, 0, 1, name1)
    my_turtle.teleport_abs_turtle(3, 7, math.pi/2, name1)
    my_turtle.set_pen_turtle(0, 0, 255, 2, name1)
    my_turtle.teleport_abs_turtle(3, 3, math.pi/2, name1)
    my_turtle.kill_turtle(name1)
    
    # Destroy node and close program
    my_turtle.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: m02_ros2_communication/m02_ros2_with_cpp/src/turtle_challenge.cpp
// ---------------------- CPP Standard Headers --------------------------------
#include <cmath>
#include <chrono>

// ---------------------- ROS2 Headers ----------------------------------------
#include <rclcpp/rclcpp.hpp>

// ---------------------- Messages and Services needed -----------------------
#include <turtlesim/msg/pose.hpp>
#include <std_srvs/srv/empty.hpp>
#include <turtlesim/srv/spawn.hpp>
#include <turtlesim/srv/set_pen.hpp>
#include <turtlesim/srv/kill.hpp>
#include <turtlesim/srv/teleport_absolute.hpp>
#include <turtlesim/srv/teleport_relative.hpp>

// ------------------- Namespaces used ---------------------------------------
using namespace std::chrono_literals;

// -------------------- Playground Class Implementation -----------------------
class PlaygroundTurtle : public rclcpp::Node
{
public:
    // User defined constructor
    PlaygroundTurtle() : Node("playground_turtle_cpp") {}

    /**
     * Call service to clear turtlesim background
     * 
     * @return true if response is received without errors.
    */
    bool clear_board()
    {
        auto request = std::make_shared<std_srvs::srv::Empty::Request>();
        while(!this->clear_client_->wait_for_service(1s))
        {
            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), 
                "Waiting for service in turtlesim: Clear");
        }
        auto result = clear_client_->async_send_request(request);
        if(rclcpp::spin_until_future_complete(this->get_node_base_interface(), result) == 
            rclcpp::FutureReturnCode::SUCCESS)
        {
            return true;
        }
        return false;
    }

    /**
     * Call service to create new turtle
     * 
     * @param x Set X position of the turtle in the turtlesim map. 
     * @param y Set Y position of the turtle in the turtlesim map.
     * @param theta Set orientation of the turtle (radians)
     * @param name Name of the turtle create
     * 
     * @return true if response is received without errors.
    */
    bool spawn_turtle(float x, float y, float theta, std::string& name)
    {
        auto request = std::make_shared<turtlesim::srv::Spawn::Request>();
        request->x = x;
        request->y = y;
        request->theta = theta;
        request->name = name;
        while(!this->spawn_client_->wait_for_service(1s))
        {
            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), 
                "Waiting for service in turtlesim: Spawn");
        }
        auto result = spawn_client_->async_send_request(request);
        if(rclcpp::spin_until_future_complete(this->get_node_base_interface(), result) == 
            rclcpp::FutureReturnCode::SUCCESS)
        {
            return true;
        }
        return false;
    }

    /**
     * Call service to delete a turtle
     * 
     * @param name Name of the turtle to delete
     * 
     * @return true if response is received without errors.
    */
    bool kill_turtle(std::string& name)
    {
        auto request = std::make_shared<turtlesim::srv::Kill::Request>();
        request->name = name;
        while(!this->kill_client_->wait_for_service(1s))
        {
            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), 
                "Waiting for service in turtlesim: Kill");
        }
        auto result = kill_client_->async_send_request(request);
        if(rclcpp::spin_until_future_complete(this->get_node_base_interface(), result) == 
            rclcpp::FutureReturnCode::SUCCESS)
        {
            return true;
        }
        return false;
    }

    /**
     * Call service to teleport the turtle (according origin of turtlesim)
     * 
     * @param x Set X position to tp turtle.
     * @param y Set Y position to tp turtle.
     * @param theta Set orientation of turtle (radians)
     * @param name Turtle to teleport
     * 
     * @return true if response is received without errors.
    */
    bool teleport_abs_turtle(float x, float y, float theta, std::string& name)
    {
        tp_abs_client_ = this->create_client<turtlesim::srv::TeleportAbsolute>
            ( "/" + name + "/teleport_absolute");
        auto request = 
            std::make_shared<turtlesim::srv::TeleportAbsolute::Request>();
        request->x = x;
        request->y = y;
        request->theta = theta;
        while(!this->tp_abs_client_->wait_for_service(1s))
        {
            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), 
                "Waiting for service in turtlesim: Teleport Absolute");
        }
        auto result = tp_abs_client_->async_send_request(request);
        if(rclcpp::spin_until_future_complete(this->get_node_base_interface(), result) == 
            rclcpp::FutureReturnCode::SUCCESS)
        {
            return true;
        }
        return false;
    }

    /**
     * Call service to teleport the turtle (veloicity of given turtle)
     * 
     * @param linear Set linear velocity of the turlte
     * @param angular Set angular velocity of the turtle
     * @param name Name of the turtle teleport.
     * 
     * @return true if response is received without errors.
    */
    bool teleport_rel_turtle(float linear, float angular, std::string& name)
    {
        tp_rel_client_ = this->create_client<turtlesim::srv::TeleportRelative>
            ( "/" + name + "/teleport_relative");
        auto request = 
            std::make_shared<turtlesim::srv::TeleportRelative::Request>();
        request->linear = linear;
        request->angular = angular;
        while(!this->tp_rel_client_->wait_for_service(1s))
        {
            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), 
                "Waiting for service in turtlesim: Teleport Relative");
        }
        auto result = tp_rel_client_->async_send_request(request);
        if(rclcpp::spin_until_future_complete(this->get_node_base_interface(), result) == 
            rclcpp::FutureReturnCode::SUCCESS)
        {
            return true;
        }
        return false;
    }

    /**
     * Call service to teleport the turtle (according origin of turtlesim)
     * 
     * @param r 0-255 intensity of red color
     * @param g 0-255 intensity of green color
     * @param b 0-255 intensity of blue color
     * @param witdh Width of the pen trance
     * @param name Name of the turtle to modify the pen configuration
     * 
     * @return true if response is received without errors.
    */
    bool set_pen_turtle(unsigned int r,unsigned int g, unsigned int b, 
        unsigned int width, std::string& name)
    {
        pen_client_ = this->create_client<turtlesim::srv::SetPen>
            ( "/" + name + "/set_pen");
        auto request = 
            std::make_shared<turtlesim::srv::SetPen::Request>();
        request->r = r;
        request->g = g;
        request->b = b;
        request->width = width;
        request->off = 0;
        while(!this->pen_client_->wait_for_service(1s))
        {
            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), 
                "Waiting for service in turtlesim: SetPen");
        }
        auto result = pen_client_->async_send_request(request);
        if(rclcpp::spin_until_future_complete(this->get_node_base_interface(), result) == 
            rclcpp::FutureReturnCode::SUCCESS)
        {
            return true;
        }
        return false;
    }

private:
    // Client definitions for turtlesim
    rclcpp::Client<std_srvs::srv::Empty>::SharedPtr clear_client_ = 
        this->create_client<std_srvs::srv::Empty>("/clear");
    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr spawn_client_ = 
        this->create_client<turtlesim::srv::Spawn>("/spawn");
    rclcpp::Client<turtlesim::srv::Kill>::SharedPtr kill_client_ = 
        this->create_client<turtlesim::srv::Kill>("/kill");
    
    // Client definition for turtles
    rclcpp::Client<turtlesim::srv::TeleportAbsolute>::SharedPtr tp_abs_client_;
    rclcpp::Client<turtlesim::srv::TeleportRelative>::SharedPtr tp_rel_client_;
    rclcpp::Client<turtlesim::srv::SetPen>::SharedPtr pen_client_;

    // Subscriptions
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_subscriber_;
};

int main(int argc, char* argv[])
{
    // Initialize ROS Client Library for C++
    rclcpp::init(argc, argv);

    // Define name and instance a playground object
    std::string name1 {"dan_turtle"};
    auto ptr_play = std::make_shared<PlaygroundTurtle>();

    // Play with the turtle
    ptr_play->clear_board();
    ptr_play->spawn_turtle(3, 3, 0, name1);
    ptr_play->teleport_abs_turtle(7, 3, 3*M_PI_2, name1);
    ptr_play->set_pen_turtle(255, 0, 0, 2, name1);
    ptr_play->teleport_abs_turtle(7, 7, -M_PI, name1);
    ptr_play->set_pen_turtle(0, 255, 0, 2, name1);
    ptr_play->teleport_abs_turtle(3, 7, M_PI_2, name1);
    ptr_play->set_pen_turtle(0, 0, 255, 2, name1);
    ptr_play->teleport_abs_turtle(3, 3, M_PI_2, name1);
    ptr_play->teleport_rel_turtle(1, M_PI_2, name1);
    ptr_play->kill_turtle(name1);
    
    // Close and shutdown
    rclcpp::shutdown();
    return 0;
}